lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,331 @@ using namespace mlir;
+
+    mapRange(ifOp.getResults(), newIfOp.getResults(), b.mapping);
+    mapRange(ifOp.thenBlock()->getArguments(),
+             newIfOp.thenBlock()->getArguments(), b.mapping);
+
+    if (ifOp.elseBlock()) {
+      mapRange(ifOp.elseBlock()->getArguments(),
+               newIfOp.elseBlock()->getArguments(), b.mapping);
+    }
+
+    b.setInsertionPointToStart(newIfOp.thenBlock());
+  }
+
+  cloneOpsInBlock(ifOp.thenBlock(), builders, schedule);
+
+  if (auto elseBlock = ifOp.elseBlock()) {
+    for (auto [builder, newIfOp] : llvm::zip(builders, newIfOps)) {
+      builder.setInsertionPointToStart(newIfOp.elseBlock());
+    }
+    cloneOpsInBlock(elseBlock, builders, schedule);
+  }
+
+  for (auto [idx, newIfOp] : llvm::zip(partitionIndices, newIfOps)) {
+    builders[idx].setInsertionPointAfter(newIfOp);
+  }
+}
+
+void cloneReduceOp(triton::ReduceOp reduceOp,
+                   SmallVector<WarpGroupBuilder> &builders,
+                   const WarpSchedule &schedule) {
+  auto partition = getPartition(reduceOp, schedule);
+  assert(partition);
+  auto partitionIndices = getPartitionIndicesToCloneInto(partition, schedule);
+
+  SmallVector<ReduceOp> newReduceOps;
+  for (size_t idx : partitionIndices) {
+    auto &b = builders[idx];
+
+    SmallVector<Value> srcs;
+    for (auto src : reduceOp.getSrcs()) {
+      srcs.push_back(b.mapping.lookupOrDefault(src));
+    }
+    auto axis = reduceOp.getAxis();
+    auto newReduceOp =
+        b.create<triton::ReduceOp>(reduceOp.getLoc(), srcs, axis);
+    newReduceOp->setAttrs(reduceOp->getAttrs());
+    newReduceOps.push_back(newReduceOp);
+
+    mapRange(reduceOp.getResults(), newReduceOp.getResults(), b.mapping);
+
+    auto &region = newReduceOp.getRegion();
+    Block *block = &region.emplaceBlock();
+    for (auto arg : reduceOp.getRegion().getBlocks().front().getArguments()) {
+      auto newArg = block->addArgument(arg.getType(), arg.getLoc());
+      b.mapping.map(arg, newArg);
+    }
+
+    b.setInsertionPointToStart(block);
+  }
+
+  cloneOpsInBlock(reduceOp.getBody(), builders, schedule);
+
+  for (auto [idx, newReduceOp] : llvm::zip(partitionIndices, newReduceOps)) {
+    builders[idx].setInsertionPointAfter(newReduceOp);
+  }
+}
+
+void cloneOpsInBlock(Block *block, SmallVector<WarpGroupBuilder> &builders,
+                     const WarpSchedule &schedule) {
+  for (auto &op_ : *block) {
+    auto op = &op_;
+    auto partition = getPartition(op, schedule);
+
+    if (auto forOp = dyn_cast<scf::ForOp>(op)) {
+      cloneForOp(forOp, builders, schedule);
+    } else if (auto ifOp = dyn_cast<scf::IfOp>(op)) {
+      cloneIfOp(ifOp, builders, schedule);
+    } else if (auto reduceOp = dyn_cast<triton::ReduceOp>(op)) {
+      cloneReduceOp(reduceOp, builders, schedule);
+    } else if (auto yieldOp = dyn_cast<scf::YieldOp>(op)) {
+      if (yieldOp.getOperands().empty()) {
+        continue;
+      }
+
+      auto doClone = [&](WarpGroupBuilder &builder) {
```

Comment:
can you use `getPartitionsToCloneInto` here to get the indices and replace the lambda + conditional with a loop over the indinces?